samza-core/src/main/java/org/apache/samza/zk/ZkDistributedReadWriteLock.java

```diff
@@ -0,0 +1,233 @@
+  private static final String PROCESSORS_PATH = "processors";
+  private final ZkUtils zkUtils;
+  private final String lockPath;
+  private final String particpantsPath;
+  private final String processorsPath;
+  private final String participantId;
+  private final ZkKeyBuilder keyBuilder;
+  private final Random random = new Random();
+  private String activeParticipantPath = null;
+  private String activeProcessorPath = null;
+  private Object mutex;
+  private Boolean isInCriticalSection = false;
+  private Boolean isStateLost = false;
+
+  public ZkDistributedReadWriteLock(String participantId, ZkUtils zkUtils, String lockId) {
+    if (zkUtils == null) {
+      throw new RuntimeException("Cannot operate ZkDistributedReadWriteLock without ZkUtils.");
+    }
+    this.zkUtils = zkUtils;
+    this.participantId = participantId;
+    this.keyBuilder = zkUtils.getKeyBuilder();
+    lockPath = String.format("%s/readWriteLock-%s", keyBuilder.getRootPath(),lockId);
+    particpantsPath = String.format("%s/%s", lockPath, PARTICIPANTS_PATH);
+    processorsPath = String.format("%s/%s", lockPath, PROCESSORS_PATH);
+    zkUtils.validatePaths(new String[] {lockPath, particpantsPath, processorsPath});
+    mutex = new Object();
+    zkUtils.getZkClient().subscribeChildChanges(particpantsPath, new ParticipantChangeHandler(zkUtils));
+    zkUtils.getZkClient().subscribeStateChanges(new ZkSessionStateChangedListener());
+  }
+
+  /**
+   * Tries to acquire a lock in order to generate run.id. On failure to acquire lock, it keeps trying until the lock times out.
+   * Creates a sequential ephemeral node under "participants" to acquire the lock.
+   * If the path of this node has the lowest sequence number,
+   * it creates a sequential ephemeral node under "processors" and checks the number of nodes under "processors"
+   * if there is only one node under "processors", a WRITE access lock is acquired
+   * else a READ access lock is acquired
+   * @param timeout Duration of lock acquiring timeout.
+   * @param unit Unit of the timeout defined above.
+   * @return AccessType.READ/WRITE if lock is acquired successfully, AccessType.NONE if it times out.
+   */
+  @Override
+  public AccessType lock(long timeout, TimeUnit unit)
+      throws TimeoutException {
+
+    activeParticipantPath = zkUtils.getZkClient().createEphemeralSequential(particpantsPath + "/", participantId);
+
+    //Start timer for timeout
+    long startTime = System.currentTimeMillis();
+    long lockTimeout = TimeUnit.MILLISECONDS.convert(timeout, unit);
+
+    while((System.currentTimeMillis() - startTime) < lockTimeout) {
+      synchronized (mutex) {
+        AccessType accessType = checkAndAcquireLock();
+        if(accessType != AccessType.NONE) {
+          isInCriticalSection = true;
+          if(isStateLost) {
+            throw new SamzaException("Lock's state lost due to connection expiry");
+          }
+          return accessType;
+        } else {
```

Comment:
when this happens?

Reply:
As in what is the scenario when this exception is thrown?

Suppose processor P1 is trying to acquire this lock. It is the first participant and the only one in the processors list and thus would have accessType = WRITE. Now, if there is a connection expiry then ZK will clean up all its ephemeral nodes meaning its participant and processor node will be deleted leading to another processor P2 acquiring the lock as WRITE. P2 would go ahead and write a run.id (say r2). Now, if we did not throw an exception in P1 and instead returned WRITE, then P1 will go ahead (upon connection restore) will overwrite the run.id (say r1). Thus P1 and P2 are using two different run.ids. It is a corner case.